clang/lib/Sema/SemaChecking.cpp

```diff
@@ -8900,7 +8900,8 @@ void Sema::CheckMemaccessArguments(const CallExpr *Call,
           << Call->getCallee()->getSourceRange());
     else if (const auto *RT = PointeeTy->getAs<RecordType>()) {
 
-      bool IsTriviallyCopyableCXXRecord =
+      bool MayBeTriviallyCopyableCXXRecord =
+          RT->isIncompleteType() ||
```

Comment:
So you're basically talking about this case:
```
struct S;

void foo(S *s1, S *s2) {
  memcpy(s1, s2, 10);
}

struct S {
  S(const S&);
  S& operator=(const S &);
};
```
I'm not certain that's worth splitting off as its own diagnostic as `memcpy`, `memset`, etc all need you to pass in a size parameter and... anything the user puts there in case of an incomplete type is pretty suspect. Based on that, I think we may *want* the diagnostic for incomplete types, but that the diagnostic only makes sense in C++ regardless of whether the type is complete or not.

However, GCC silences the diagnostic for incomplete types: https://godbolt.org/z/bsPnrKf19 so I may be missing some use case where this construct is safer than I'm thinking.

Reply:
According to https://godbolt.org/z/xcnb154de having fields that point to incomplete types does not make you non trivially copyable, so I'm not even sure it reduces the quality of the original warning.

Reply:
> According to https://godbolt.org/z/xcnb154de having fields that point to incomplete types does not make you non trivially copyable, so I'm not even sure it reduces the quality of the original warning.

Why WOULD it?  Those are pointers, which are complete types.

> So you're basically talking about this case:
> 
> ```
> struct S;
> 
> void foo(S *s1, S *s2) {
>   memcpy(s1, s2, 10);
> }
> 
> struct S {
>   S(const S&);
>   S& operator=(const S &);
> };
> ```
> 
> I'm not certain that's worth splitting off as its own diagnostic as `memcpy`, `memset`, etc all need you to pass in a size parameter and... anything the user puts there in case of an incomplete type is pretty suspect. Based on that, I think we may _want_ the diagnostic for incomplete types, but that the diagnostic only makes sense in C++ regardless of whether the type is complete or not.
> 
> However, GCC silences the diagnostic for incomplete types: https://godbolt.org/z/bsPnrKf19 so I may be missing some use case where this construct is safer than I'm thinking.

Yeah, something like that.  It ends up causing someone to be able to 'side-step' this diagnostic inadvertently by using an incomplete type.  So a separate warning of, "you're using an incomplete type here that might be non-trivially-copyable in the future ya know!" might be useful to some, though if we put it in the same group as this one, it might cause folks to throw the baby with the bath water.

More 'mulling' than having a request.

Reply:
Trivial copyability is about whether you can copy the object itself (aka, its members), and you can't create an object of incomplete class type or which contains an object of incomplete class type. So a pointer to an incomplete type is fine in that regard; it's just a pointer, you're not dereferencing it. However, for this diagnostic, `memcpy` and friends are touching the bytes of the object (so it's a dereference, effectively), and passing a pointer to an incomplete type means those are touching bytes of an unknown type which may be completed later.

Reply:
Should I mention that as a comment in the code?

Reply:
> Should I mention that as a comment in the code?

There were a few things, what do you mean by 'that'?  Perhaps something that "note, this makes the diagnostic imperfect because... but *shrug*"?

Reply:
Something along those lines?

```
FIXME: Do not consider incomplete type even though they may be completed later.
```

Reply:
I can live with that.

Reply:
> Something along those lines?
> 
> ```
> FIXME: Do not consider incomplete type even though they may be completed later.
> ```

I'd prefer something more like:
```
FIXME: Do not consider incomplete types even though they may be completed later. GCC does
not diagnose such code, but we may want to consider diagnosing it in the future, perhaps under
a different, but related, diagnostic group.
```
so it's more clear why it's a fixme. But yeah, I can live with that.